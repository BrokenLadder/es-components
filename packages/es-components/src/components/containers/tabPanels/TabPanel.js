import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider } from 'styled-components';
import defaultTheme from 'es-components-via-theme';
import Tab from './Tab';

const TabWrapper = styled.div`
  display: flex;
  border-bottom: 1px solid ${props => props.theme.colors.gray4};
`;

const TabFormatter = styled.div`
  display: flex;
  flex-direction: column;
  font-size: inherit;
  @media (max-width: ${props => props.theme.screenSize.desktop}) {
    width: 100%;
  }
  @media (min-width: ${props => props.theme.screenSize.desktop}) {
    flex-direction: row;
    justify-content: flex-start;
  }
`;

const AriaAnnouncer = styled.p`
  position: fixed;
  color: transparent;
  left: -1000px;
  top: 10px;
`;

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    const child = Array.isArray(this.props.children)
      ? props.children[0]
      : props.children;
    this.state = {
      value: child.props.name,
      currentContent: child.props.children,
      simpleName: child.props.simpleName || child.props.name
    };
    this.tabChanged = this.tabChanged.bind(this);
  }

  tabChanged(name, child, simpleName) {
    this.setState({
      value: name,
      currentContent: child,
      simpleName: simpleName || name
    });
  }

  render() {
    const { theme, children } = this.props;
    const elements = React.Children.map(children, (child, i) => {
      const isSelected = child.props.name === this.state.value;
      return React.cloneElement(child, {
        key: child.props.name,
        selected: isSelected,
        action: this.tabChanged,
        selectedName: this.state.value,
        theme
      });
    });

    return (
      <ThemeProvider theme={theme}>
        <div>
          <AriaAnnouncer id="announcer" aria-live="assertive">{`${
            this.state.simpleName
          } Sub text is now showing`}</AriaAnnouncer>
          <TabWrapper>
            <TabFormatter>{elements}</TabFormatter>
          </TabWrapper>
          <div>{this.state.currentContent}</div>
        </div>
      </ThemeProvider>
    );
  }
}

function childrenRule(props, propName, component) {
  let children = props[propName];
  if (!Array.isArray(children)) {
    children = [children];
  }
  if (!children.every(child => child.type.name === 'Tab')) {
    return new Error('Tab Panel only accepts Tabs as direct descendants.');
  }

  return null;
}

TabPanel.propTypes = {
  /**
   * Theme to be applied to the tab panel. Can be used to style the Tab and Tab List
   */
  theme: PropTypes.object,
  /**
   * Makes sure immediate children are Tab or Tab List, as we cannot render anything else in the tab heading.
   */
  children: childrenRule
};

TabPanel.defaultProps = {
  theme: defaultTheme
};

TabPanel.Tab = Tab;

export default TabPanel;
