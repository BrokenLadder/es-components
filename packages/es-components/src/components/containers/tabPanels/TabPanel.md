Tab Panels allow us to store lots more data on the screen and only show the amount we want, when we want. This works by having a Tab contain context, or a Tab List that contains elements. Each direct child of the tab list must contain optiontext that will turn into the title in the dropdown. Tabs are able to have a full node inside instead of just text. This allows us to put in icons and other styling.

```
<div>
<TabPanel>
  <TabPanel.Tab name="Hi there">
    <p>
      HELLO WORLD!!!!
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab simpleName="interesting" name={<div><Icon name="certificate" />Interesting 5 plans</div>}>
    <div>
      <p>
        Nesting is fun
      </p>
    </div>
  </TabPanel.Tab>
  <TabPanel.Tab name="Lorem ipsum">
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab name="dolor sit">
    <p>
      ed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?
    </p>
  </TabPanel.Tab>
</TabPanel>
</div>
```