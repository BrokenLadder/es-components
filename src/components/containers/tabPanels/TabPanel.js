import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider } from 'styled-components';
import viaTheme from '../../theme/viaTheme';
import TabList from './TabList';
import Tab from './Tab';

const TabWrapper = styled('div')`
  display: flex;
  justify-content: center;
  font-size: 24px;

  @media (min-width: 899px) {
    justify-content: flex-start;
  }
`;

const TabFormatter = styled('div')`
  display: flex;
  justify-content: center;
  flex-direction: column;

  @media (min-width: 899px) {
    flex-direction: row;
  }
`;

const TabContent = styled('div')`
  border-top: 1px solid #ddd;
`;

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      currentContent: ''
    };
    this.tabChanged = this.tabChanged.bind(this);
  }

  componentWillMount() {
    if (this.state.value === '') {
      const child = this.props.children[0];
      const selectedChild = this.getFirstValueChild(child);
      this.tabChanged(
        selectedChild.props.name ||
          selectedChild.props.optionText.replace(/\s/g, ''),
        selectedChild.props.children
      );
    }
  }

  getFirstValueChild(child) {
    if (Array.isArray(child.props.children)) {
      return child.props.children[0];
    }
    return child;
  }

  listIsSelected(list) {
    if (Array.isArray(list.props.children)) {
      return (
        list.props.children.filter(child => {
          if (child.props.optionText) {
            return (
              child.props.optionText.replace(/\s/g, '') === this.state.value
            );
          }
          return false;
        }).length > 0
      );
    }
    return false;
  }

  tabChanged(name, child) {
    this.setState({ value: name, currentContent: child });
  }

  render() {
    const { theme, children } = this.props;
    const elements = React.Children.map(children, (child, i) => {
      const isSelected =
        child.props.name === this.state.value ||
        this.listIsSelected(child) ||
        (i === 0 && this.state.value === '');
      return React.cloneElement(child, {
        key: child.props.name,
        selected: isSelected,
        action: this.tabChanged,
        selectedName: this.state.value,
        theme
      });
    });

    return (
      <ThemeProvider theme={theme}>
        <div>
          <TabWrapper>
            <TabFormatter>{elements}</TabFormatter>
          </TabWrapper>
          <TabContent>{this.state.currentContent}</TabContent>
        </div>
      </ThemeProvider>
    );
  }
}

TabPanel.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
};

TabPanel.defaultProps = {
  theme: viaTheme
};

TabPanel.TabList = TabList;
TabPanel.Tab = Tab;

export default TabPanel;
