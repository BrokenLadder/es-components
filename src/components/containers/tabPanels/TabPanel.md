Test Tab Panels

```
<div>
<TabPanel optionKeyFunc={((x) => x.replace(/\s/g, ''))}>
  <TabPanel.TabList name="testList">
    <span optiontext="test3">
       No elements
    </span>
    <span optiontext="test4">
       <p>
        More information. So much information. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
       </p>
    </span>
  </TabPanel.TabList>
  <TabPanel.Tab name="test1">
    <p>
      HELLO WORLD!!!!
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab name="test2">
    <div>
      <p>
        Nesting is fun
      </p>
    </div>
  </TabPanel.Tab>
</TabPanel>
</div>
```