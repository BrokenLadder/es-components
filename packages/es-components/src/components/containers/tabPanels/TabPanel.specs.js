/* eslint-env jest */

import React from 'react';
import { mountWithTheme, renderWithTheme } from 'styled-enzyme';

import TabPanel from './TabPanel';

describe('Tab panel component', () => {
  const instanceToRender = (
    <TabPanel optionKeyFunc={x => x.replace(/\s/g, '')}>
      <TabPanel.Tab name="test1">
        <div>Test Content 1</div>
      </TabPanel.Tab>
      <TabPanel.Tab name="test2">
        <div>Test Content 2</div>
      </TabPanel.Tab>
    </TabPanel>
  );

  const instanceWithAnnouncer = (
    <TabPanel optionKeyFunc={x => x.replace(/\s/g, '')}>
      <TabPanel.Tab
        name="test1"
        simpleName="I'm Simple Name!"
        announcerText="Why am I not just part of simpleName?"
      >
        <div>Test Content 1</div>
      </TabPanel.Tab>
      <TabPanel.Tab name="test2">
        <div>Test Content 2</div>
      </TabPanel.Tab>
    </TabPanel>
  );

  it('renders as expected', () => {
    const tree = renderWithTheme(instanceToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders as expected with aria announcer text', () => {
    const tree = renderWithTheme(instanceWithAnnouncer).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Will select first child on load', () => {
    const tabPanelInstance = mountWithTheme(instanceToRender);
    expect(tabPanelInstance.state('value')).toBe('test1');
  });

  it('Properly selects a tab', () => {
    const tabPanelInstance = mountWithTheme(instanceToRender);
    tabPanelInstance.setState({ value: 'test2' });
    expect(
      tabPanelInstance
        .find({ name: 'test2' })
        .first()
        .props().selected
    ).toBe(true);
  });

  it('Clicks on tab select them', () => {
    const tabPanelInstance = mountWithTheme(instanceToRender);
    const tab = tabPanelInstance.find({ name: 'test2' });
    const tabButton = tab.find('button');
    tabButton.simulate('click');
    expect(tabPanelInstance.state('value')).toBe('test2');
  });

  it('Clicking on tab invokes the callback', () => {
    let selectedTab = 'test1';
    const tabPanelInstance = mountWithTheme(instanceToRender);
    tabPanelInstance.setProps({
      tabChanged(name) {
        selectedTab = name;
      }
    });

    const tab = tabPanelInstance.find({ name: 'test2' });
    const tabButton = tab.find('button');
    tabButton.simulate('click');
    expect(selectedTab).toBe('test2');
  });

  it('Clicking the selected tab does not invoke the callback unnecessarily', () => {
    let timesCalled = 0;
    const tabPanelInstance = mountWithTheme(instanceToRender);
    tabPanelInstance.setProps({
      tabChanged() {
        timesCalled += 1;
      }
    });

    const tab = tabPanelInstance.find({ name: 'test2' });
    const tabButton = tab.find('button');
    tabButton.simulate('click');
    tabButton.simulate('click');
    expect(timesCalled).toBe(1);
  });
});
