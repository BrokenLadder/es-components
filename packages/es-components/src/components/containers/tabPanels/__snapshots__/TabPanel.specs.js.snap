// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab panel component renders as expected 1`] = `
<div
  className="es-tab-panel"
>
  <div
    className="es-tab-panel__wrapper sc-htpNat cixthU"
  >
    <div
      className="es-tab-panel__tabs sc-bxivhb gWEfbh"
    >
      <button
        aria-expanded={true}
        aria-label="test1 tab"
        className="sc-bdVaJa jjYEoD"
        onClick={[Function]}
        selected={true}
      >
        test1
      </button>
      <button
        aria-expanded={true}
        aria-label="test2 tab"
        className="sc-bdVaJa jjYEoD"
        onClick={[Function]}
        selected={true}
      >
        test2
      </button>
    </div>
  </div>
  <div
    className="es-tab-panel__content sc-ifAKCX MuRpR"
  >
    <div>
      Test Content 1
    </div>
  </div>
</div>
`;

exports[`Tab panel component renders as expected with aria announcer text 1`] = `
<div
  className="es-tab-panel"
>
  <div
    className="es-tab-panel__wrapper sc-htpNat cixthU"
  >
    <div
      className="es-tab-panel__tabs sc-bxivhb gWEfbh"
    >
      <button
        aria-expanded={true}
        aria-label="I'm Simple Name! tab"
        className="sc-bdVaJa jjYEoD"
        onClick={[Function]}
        selected={true}
      >
        <p
          aria-live="assertive"
          className="sc-bwzfXH iOBbZz"
          id="announcer"
        >
          I'm Simple Name! Why am I not just part of simpleName?
        </p>
        test1
      </button>
      <button
        aria-expanded={true}
        aria-label="test2 tab"
        className="sc-bdVaJa jjYEoD"
        onClick={[Function]}
        selected={true}
      >
        test2
      </button>
    </div>
  </div>
  <div
    className="es-tab-panel__content sc-ifAKCX MuRpR"
  >
    <div>
      Test Content 1
    </div>
  </div>
</div>
`;
